### PR TITLE
Prefix PRNG Part 2: implement 1D prefix PRNG

### PR DESCRIPTION
**Pull request by j-towns**
This solves https://github.com/google/jax/issues/2155 in the 1 dimensional case, so after this change, with the flag `JAX_ENABLE_PREFIX_PRNG=True`, we have
```python
In [1]: from jax import random

In [2]: k = random.PRNGKey(0)

In [3]: random.uniform(k, (3,))
Out[3]: DeviceArray([0.41845703, 0.60049903, 0.21629536], dtype=float32)

In [4]: random.uniform(k, (5,))
Out[4]: DeviceArray([0.41845703, 0.60049903, 0.21629536, 0.8041241 , 0.9653214 ], dtype=float32)
```
as required. With the default `JAX_ENABLE_PREFIX_PRNG=False` the PRNG is unaffected by this pr.

I believe the general case can also be solved now using [this idea](https://github.com/google/jax/issues/2155#issuecomment-582515875), but I'll do that in a follow up.

### CODE REVIEW

File: jax/random.py [hunk 1]
```diff
@@ -215,17 +218,43 @@ def threefry_2x32(keypair, count):
...
 
-  x = threefry2x32_p.bind(key1, key2, x[0], x[1])
-  out = np.concatenate(x)
+  out = np.stack(threefry2x32_p.bind(key1, key2, count_high, count_low), -1)
   assert out.dtype == onp.uint32
-  return lax.reshape(out[:-1] if odd_size else out, count.shape)
 
+  return out
+
+def _interleave(arr_1, arr_2):
```
Comment (j-towns) [inline jax/random.py]: Could this interleaving, which replaces what was previously concatenation, cause a performance regression?

**Comment by j-towns** [comment]
Closing temporarily cos I realise there are a few tests still to fix.

**Comment by j-towns** [comment]
Reopening because the tests should now pass.

File: jax/random.py [hunk 2]
```diff
@@ -200,32 +201,63 @@ def _broadcast(x):
...
+  out = ops.index_update(out, slice(1, None, 2), arr_2)
+  return out
+
+def _make_count(length):
+  """Produce a shape `(length, 2)` array, with pairs of uint32 elements
+  simulating a single uint64 counter.
+  """
+  if length > (1 << 32):
+    # This maximum could be raised, but need to fix
+    # https://github.com/google/jax/issues/2178 first.
```
Comment (j-towns) [inline jax/random.py]: Looks like this is being fixed already by https://github.com/google/jax/pull/2182!

**Comment by j-towns** [comment]
Following discussion with @mattjj we agreed that this should be behind a flag for now to avoid issues with reproducibility. Will implement that later.

File: jax/random.py [hunk 3]
```diff
@@ -200,32 +207,51 @@ def _broadcast(x):
       _threefry2x32_gpu_translation_rule
 
 @jit
+@partial(vmap, in_axes=(None, 0))
```
Comment (mattjj) [inline jax/random.py]: Instead of using `vmap` here, can we just rely on the (elementwise) implementation broadcasting along leading axes? That is, is the `vmap` necessary (even if we tweak the `np.stack` calls or something)? As written with `vmap`, it means the function is no longer rank-polymorphic IIUC, and has a bit more complexity.

File: jax/random.py [hunk 4]
```diff
@@ -200,32 +207,51 @@ def _broadcast(x):
...
-  if odd_size:
-    x = list(np.split(np.concatenate([count.ravel(), onp.uint32([0])]), 2))
-  else:
-    x = list(np.split(count.ravel(), 2))
+  assert np.shape(keypair) == (2,)
+  key1, key2 = keypair
 
-  x = threefry2x32_p.bind(key1, key2, x[0], x[1])
-  out = np.concatenate(x)
+  assert count.shape == (2,)
```
Comment (mattjj) [inline jax/random.py]: This could be an assertion that the trailing axis is size 2.

File: jax/random.py [hunk 5]
```diff
@@ -200,32 +207,51 @@ def _broadcast(x):
...
+  count_high, count_low = count
+
+  # Result of theefry2x32_p is a list
+  out = np.asarray(threefry2x32_p.bind(key1, key2, count_high, count_low))
```
Comment (mattjj) [inline jax/random.py]: I think using concatenate or stack is more explicit (and perhaps faster) than `np.asarray`.

File: jax/random.py [hunk 6]
```diff
@@ -200,32 +207,51 @@ def _broadcast(x):
...
   assert out.dtype == onp.uint32
-  return lax.reshape(out[:-1] if odd_size else out, count.shape)
 
+  return out
+
+def _make_count(key, at_least_size):
```
Comment (mattjj) [inline jax/random.py]: Comment to explain why this is abstracted out? IIUC it's basically because of the flag.

File: jax/random.py [hunk 7]
```diff
@@ -44,6 +46,11 @@
 from jax.interpreters import xla
 from jax.util import prod
 
+FLAGS = flags.FLAGS
+flags.DEFINE_bool('jax_enable_masking_prng',
+                  bool_env('JAX_ENABLE_MASKING_PRNG', False),
```
Comment (mattjj) [inline jax/random.py]: Can we rename this flag not to be about masking, perhaps just call it "prefix_prng"?

File: jax/random.py [hunk 8]
```diff
@@ -261,8 +291,8 @@ def fold_in(key, data):
 
 @jit
 def _fold_in(key, data):
-  key2 = lax.tie_in(key, PRNGKey(data))
-  return threefry_2x32(key, key2)
+  key2 = lax.tie_in(key, np.asarray([[0, data]], dtype='uint32'))
```
Comment (mattjj) [inline jax/random.py]: Instead of always padding left with zeros, we could make this function handle 64bit values like you do elsewhere. Maybe leave a TODO?

I think concatenating a zero with lax operations would be better than using `np.asarray` here, since `np.asarray` can be much more complex. Under a jit it's okay, but it's kind of a bad style for others to copy.

File: jax/random.py [hunk 9]
```diff
@@ -261,8 +291,8 @@ def fold_in(key, data):
...
+  return threefry_2x32(key, key2)[0]
```
Comment (mattjj) [inline jax/random.py]: I'm surprised at this indexing; it seems like the old version returned the full output of `threefry_2x32`. What changed? Did this change the type of the function?

Comment (j-towns) [reply jax/random.py]: This is just because threefry now returns a shape (1, 2) array where before it returned shape (2,). I could use a reshape/ravel or something to get the same effect here.

Comment (j-towns) [reply jax/random.py]: Maybe getting rid of vamp (re-enabling rank polymorphism) would be a better solution 🙂

File: jax/random.py (hunk 8 above)
Comment (j-towns) [reply jax/random.py]: I've reverted to feeding `data` into `PRNGKey`, so that higher order bits are handled, as they were before (I got thrown by the docstring which previously said that fold_in only handled 32 bit ints).

File: jax/random.py (hunk 6 above)
Comment (j-towns) [reply jax/random.py]: Yeah, it's basically to do flag handling in the two places where it's necessary (`_random_bits` and `_split`).

**Comment by j-towns** [comment]
There are two tests failing on travis. `tests/lax_test.py::LaxAutodiffTest::testConvGeneralDilatedGrad_lhs_shape=float16[4,2,6,7]_rhs_shape=float16[2,2,1,2]_strides=(1, 1)_padding=((1, 0), (0, 1))_lhs_dilation=(1, 1)_rhs_dilation=(2, 2)_dims=NHWC,OIHW,NCHW_feature_group_count=1_batch_group_count=2` is also failing on master so that's not caused by this pr. The other one is only failing on the py37 doctest configuation:

```
Exception occurred:
  File "/home/travis/build/j-towns/jax/jax/config.py", line 70, in check_exists
    raise Exception("Unrecognized config option: {}".format(name))
Exception: Unrecognized config option: __sphinx_mock__
```

Will see if I can figure out the cause.

**Comment by hawkinsp** [comment]
Closing as stale.
